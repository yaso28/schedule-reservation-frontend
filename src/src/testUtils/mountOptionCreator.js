import { createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import routes from '@/router/routes'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import i18nOptions from '@/i18n/options'
import VueMeta from 'vue-meta'
import CoreuiVue from '@coreui/vue'
import icons from '@/icons'
import services from '@/services'

class MountOptionCreator {
  create (storeModules = null) {
    const localVue = createLocalVue()

    // router
    localVue.use(VueRouter)
    const router = new VueRouter({
      mode: 'history',
      base: process.env.BASE_URL,
      routes
    })

    // i18n
    localVue.use(VueI18n)
    const i18n = new VueI18n(i18nOptions)

    // vue-meta
    localVue.use(VueMeta)

    // Core UI
    localVue.use(CoreuiVue)

    const option = {
      localVue,
      router,
      i18n,
      icons,
      provide: services
    }

    // store
    if (storeModules) {
      localVue.use(Vuex)
      option.store = new Vuex.Store({
        modules: storeModules
      })
    }

    return option
  }
}

export default new MountOptionCreator()
