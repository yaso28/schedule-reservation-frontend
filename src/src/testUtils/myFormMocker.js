import MyForm from '@/models/MyForm'
jest.mock('@/models/MyForm')

class MyFormMocker {
  createMockSingleField (field, initValue) {
    const input = {}
    input[field] = initValue
    return this.createMock(input)
  }

  createMock (input) {
    const mockForm = new MyForm(input)
    mockForm.values = input
    mockForm.errors = {}
    return mockForm
  }

  async setValue (wrapper, form, field, value) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        values: Object.assign({}, form.values, { [field]: value })
      })
    })
  }

  setValueNonProps (form, field, value) {
    form.values[field] = value
  }

  async setError (wrapper, form, field, errorMessage) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        errors: Object.assign({}, form.errors, { [field]: errorMessage })
      })
    })
  }
}

export default new MyFormMocker()
