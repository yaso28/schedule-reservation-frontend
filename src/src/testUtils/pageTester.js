import { mount } from '@vue/test-utils'
import MyPage from '@/components/containers/MyPage'

class PageTester {
  async mount (pageComponent, route, mountOption) {
    await mountOption.router.push(route)
    const pageWrapper = mount(MyPage, mountOption)
    return pageWrapper.findComponent(pageComponent)
  }

  assertTitleTag () {
    expect(document.title).toMatchSnapshot()
  }

  assertMetaTags () {
    expect(document.querySelectorAll('meta[data-vue-meta]')).toMatchSnapshot()
  }
}

export default new PageTester()
