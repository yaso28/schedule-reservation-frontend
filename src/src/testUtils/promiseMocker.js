import _ from 'lodash'

const createSetTimeoutPromise = getResult => new Promise((resolve, reject) => {
  setTimeout(() => getResult(resolve, reject), 1000)
})
const createSwitchPromise = (value, error, switchObj) => createSetTimeoutPromise(
  (resolve, reject) => switchObj.isSuccess ? resolve(value) : reject(error)
)
const createResolvedPromise = value => createSetTimeoutPromise(resolve => resolve(value))
const createRejectedPromise = error => createSetTimeoutPromise((resolve, reject) => reject(error))

const createMock = createPromise => jest.fn().mockImplementationOnce(createPromise)
const createMockForCases = (caseList, createPromiseForCurrentCase) => jest.fn((...args) => {
  for (let i = 0; i < caseList.length; i++) {
    const currentCase = caseList[i]
    if (_.isEqual(args, currentCase.args)) {
      return createPromiseForCurrentCase(currentCase)
    }
  }
  throw new Error('createMockForCases: no args matched')
})

class PromiseMocker {
  init () {
    jest.useFakeTimers()
  }

  clear () {
    jest.clearAllTimers()
  }

  flushPendings () {
    jest.runOnlyPendingTimers()
  }

  createSwitch () {
    return { isSuccess: true }
  }

  setSwitch (switchObj, isSuccess) {
    switchObj.isSuccess = isSuccess
  }

  mock (value, error, switchObj) {
    return createMock(() => createSwitchPromise(value, error, switchObj))
  }

  mockResolved (value) {
    return createMock(() => createResolvedPromise(value))
  }

  mockRejected (error) {
    return createMock(() => createRejectedPromise(error))
  }

  mockForCases (caseList) {
    return createMockForCases(caseList, currentCase => createSwitchPromise(currentCase.value, currentCase.error, currentCase.switch))
  }

  mockResolvedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createResolvedPromise(currentCase.value))
  }

  mockRejectedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createRejectedPromise(currentCase.error))
  }
}

export default new PromiseMocker()
