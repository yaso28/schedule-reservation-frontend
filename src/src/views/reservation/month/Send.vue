<template>
  <div>
    <slot name="h1" />

    <CForm
      v-if="loaded"
      @submit.prevent="showConfirm = true"
    >
      <CCard>
        <CCardBody>
          <dl class="row">
            <dt class="col-sm-3">
              {{ $t('month.model') }}
            </dt>
            <dd class="col-sm-9">
              {{ month.name }}
            </dd>
            <dt class="col-sm-3">
              {{ $t('reservation_status.model') }}
            </dt>
            <dd class="col-sm-9">
              {{ month.reservation_status.name }}
            </dd>
            <dt class="col-sm-3">
              {{ $t('schedule_status.model') }}
            </dt>
            <dd class="col-sm-9">
              {{ month.schedule_status.name }}
            </dd>
          </dl>
        </CCardBody>
      </CCard>

      <CCard>
        <CCardBody>
          <MyText
            :label="$t('send.mail_to')"
            :form="form"
            field="mail_to"
          />
          <MyText
            :label="$t('send.subject')"
            :form="form"
            field="subject"
          />
          <MyTextarea
            :label="$t('send.message')"
            :form="form"
            field="message"
            :rows="40"
          />

          <ul>
            <li>
              {{ $t('send.statement.setting') }}<CLink :to="{name: 'reservation.setting.edit'}">
                {{ $t('common.here') }}
              </CLink>
            </li>
            <li>
              {{ $t('send.statement.demo') }}
            </li>
          </ul>
        </CCardBody>
        <CCardFooter>
          <CButton
            class="btn-reset"
            color="secondary"
            @click="form.reset()"
          >
            {{ $t("action.reset") }}
          </CButton>
          <CButton
            type="submit"
            color="primary"
          >
            {{ $t('action.send') }}
          </CButton>
        </CCardFooter>
      </CCard>
    </CForm>

    <CModal
      :show.sync="showConfirm"
      color="primary"
      centered
      size="sm"
      :title="$t('message.send.confirm')"
      @update:show="onModalShowChange"
    >
      <template #body-wrapper>
        <div />
      </template>
    </CModal>
  </div>
</template>

<script>
export default {
  name: 'MonthSend',
  components: {
    MyText: require('@/components/inputs/MyText').default,
    MyTextarea: require('@/components/inputs/MyTextarea').default
  },
  inheritAttrs: false,
  inject: ['apiService'],
  props: {
    id: {
      type: [Number, String],
      required: true
    }
  },
  data () {
    return {
      month: null,
      form: null,
      showConfirm: false
    }
  },
  computed: {
    loaded () {
      return this.month && this.form
    }
  },
  watch: {
    id () {
      this.load()
    }
  },
  mounted () {
    this.load()
  },
  methods: {
    async load () {
      try {
        this.$store.commit('processing/beginGet')
        const info = await this.apiService.prepareMonthScheduleSendInfo(this.id)
        this.month = info.month
        this.form = this.apiService.makeForm({
          mail_to: info.mail_to,
          subject: info.subject,
          message: info.message
        })
      } catch (error) {
        if (!this.apiService.isHandledError(error)) {
          throw error
        }
      } finally {
        this.$store.commit('processing/endGet')
      }
    },
    onModalShowChange (isShow, event, isOk) {
      if (isOk) {
        this.send()
      }
    },
    async send () {
      try {
        this.$store.commit('processing/beginPost')
        this.$store.commit('alert/clear')
        await this.apiService.sendMonthSchedule(this.id, this.form)
        this.$store.dispatch('alert/sendSuccess')
        this.$router.push({ name: 'reservation.month.list' })
      } catch (error) {
        if (!this.apiService.isHandledError(error)) {
          throw error
        }
      } finally {
        this.$store.commit('processing/endPost')
      }
    }
  }
}
</script>
