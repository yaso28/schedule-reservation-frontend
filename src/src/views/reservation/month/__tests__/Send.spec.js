import Send from '../Send'
import mountOptionCreator from '@/testUtils/mountOptionCreator'
import pageTester from '@/testUtils/pageTester'
import promiseMocker from '@/testUtils/promiseMocker'
import storeModules from '@/store/modules'
import myFormMocker from '@/testUtils/myFormMocker'
import { CModal } from '@coreui/vue'

jest.mock('@/store/modules')
jest.mock('@/models/MyForm')

const sendInfo = require('@/testUtils/dummyApiData/month/send-prepare.json')
const id = sendInfo.month.id
const makeFormArgs = {
  mail_to: sendInfo.mail_to,
  subject: sendInfo.subject,
  message: sendInfo.message
}
const sendResult = require('@/testUtils/dummyApiData/common/update.json')
sendResult.id = id

const getModalWrapper = wrapper => wrapper.findComponent(CModal)
const chooseModal = function (wrapper, isOk) {
  wrapper.findComponent(CModal).vm.$emit('update:show', false, null, isOk)
}

describe('Send', () => {
  let mountOption
  let apiService
  beforeEach(() => {
    promiseMocker.init()
    mountOption = mountOptionCreator.create(storeModules)
    apiService = mountOption.provide.apiService
  })

  afterEach(() => {
    promiseMocker.clear()
    jest.resetAllMocks()
  })

  describe('マウント', () => {
    let wrapper
    let loadError
    let loadSwitch
    beforeEach(async () => {
      loadError = new Error('load')
      loadSwitch = promiseMocker.createSwitch()
      apiService.prepareMonthScheduleSendInfo = promiseMocker.mock(sendInfo, loadError, loadSwitch)
      wrapper = await pageTester.mount(Send, { name: 'reservation.month.send', params: { id } }, mountOption)
    })

    test('titleタグ', () => {
      pageTester.assertTitleTag()
    })

    test('metaタグ', () => {
      pageTester.assertMetaTags()
    })

    test('Processingカバー表示開始', () => {
      expect(storeModules.processing.mutations.beginGet).toHaveBeenCalled()
    })

    test('データ取得処理を呼出', () => {
      expect(apiService.prepareMonthScheduleSendInfo).toHaveBeenCalledWith(id)
    })

    test('Processingカバー表示中', () => {
      expect(storeModules.processing.mutations.endGet).not.toHaveBeenCalled()
    })

    describe('データ取得：エラー', () => {
      beforeEach(() => {
        promiseMocker.setSwitch(loadSwitch, false)
        apiService.isHandledError = jest.fn().mockReturnValueOnce(true)
        promiseMocker.flushPendings()
      })

      test('エラー処理', () => {
        expect(apiService.isHandledError).toHaveBeenCalledWith(loadError)
      })

      test('Processingカバー表示終了', () => {
        expect(storeModules.processing.mutations.endGet).toHaveBeenCalled()
      })
    })

    describe('データ取得：成功', () => {
      let form
      beforeEach(() => {
        form = myFormMocker.createMock(makeFormArgs)
        apiService.makeForm = jest.fn().mockReturnValueOnce(form)
        promiseMocker.flushPendings()
      })

      test('フォーム作成', () => {
        expect(apiService.makeForm).toHaveBeenCalledWith(makeFormArgs)
      })

      test('ページ全体表示', () => {
        expect(wrapper.element).toMatchSnapshot()
      })

      test('Processingカバー表示終了', () => {
        expect(storeModules.processing.mutations.endGet).toHaveBeenCalled()
      })

      describe('URL変更', () => {
        let newId
        beforeEach(() => {
          newId = id + 1
          apiService.prepareMonthScheduleSendInfo = promiseMocker.mockResolved(sendInfo)
          storeModules.processing.mutations.beginGet.mockReset()
          storeModules.processing.mutations.endGet.mockReset()
          wrapper.setProps({ id: newId })
        })

        test('Processingカバー表示開始', () => {
          expect(storeModules.processing.mutations.beginGet).toHaveBeenCalled()
        })

        test('データ取得処理を呼出', () => {
          expect(apiService.prepareMonthScheduleSendInfo).toHaveBeenCalledWith(newId)
        })

        test('Processingカバー表示中', () => {
          expect(storeModules.processing.mutations.endGet).not.toHaveBeenCalled()
        })
      })

      describe('リセットボタンクリック', () => {
        beforeEach(() => {
          wrapper.find('button.btn-reset').trigger('click')
        })

        test('フォームリセット', () => {
          expect(form.reset).toHaveBeenCalledWith()
        })
      })

      describe('フォーム送信', () => {
        let sendError
        let sendSwitch
        beforeEach(async () => {
          sendError = new Error('send')
          sendSwitch = promiseMocker.createSwitch()
          apiService.sendMonthSchedule = promiseMocker.mock(sendResult, sendError, sendSwitch)
          await wrapper.find('form').trigger('submit')
        })

        test('確認モーダル表示', () => {
          expect(getModalWrapper(wrapper).props().show).toBeTruthy()
        })

        describe('確認モーダル：キャンセル', () => {
          beforeEach(() => {
            chooseModal(wrapper, false)
          })

          test('Processingカバーは表示しない', () => {
            expect(storeModules.processing.mutations.beginPost).not.toHaveBeenCalled()
          })

          test('アラートはクリアしない', () => {
            expect(storeModules.alert.mutations.clear).not.toHaveBeenCalled()
          })

          test('送信処理は呼び出さない', () => {
            expect(apiService.sendMonthSchedule).not.toHaveBeenCalled()
          })
        })

        describe('確認モーダル：OK', () => {
          let router
          beforeEach(() => {
            router = mountOption.router
            router.push = jest.fn()
            chooseModal(wrapper, true)
          })

          test('Processingカバー表示開始', () => {
            expect(storeModules.processing.mutations.beginPost).toHaveBeenCalled()
          })

          test('アラートクリア', () => {
            expect(storeModules.alert.mutations.clear).toHaveBeenCalled()
          })

          test('送信処理を呼出', () => {
            expect(apiService.sendMonthSchedule).toHaveBeenCalledWith(id, form)
          })

          test('Processingカバー表示中', () => {
            expect(storeModules.processing.mutations.endPost).not.toHaveBeenCalled()
          })

          describe('送信：エラー', () => {
            beforeEach(() => {
              promiseMocker.setSwitch(sendSwitch, false)
              apiService.isHandledError = jest.fn().mockReturnValueOnce(true)
              promiseMocker.flushPendings()
            })

            test('エラー処理', () => {
              expect(apiService.isHandledError).toHaveBeenCalledWith(sendError)
            })

            test('Processingカバー表示終了', () => {
              expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
            })

            test('成功アラートは呼び出さない', () => {
              expect(storeModules.alert.actions.sendSuccess).not.toHaveBeenCalled()
            })

            test('リダイレクトしない', () => {
              expect(router.push).not.toHaveBeenCalled()
            })
          })

          describe('保存：成功', () => {
            beforeEach(() => {
              promiseMocker.flushPendings()
            })

            test('成功アラート', () => {
              expect(storeModules.alert.actions.sendSuccess).toHaveBeenCalled()
            })

            test('リダイレクト', () => {
              expect(router.push).toHaveBeenCalledWith({ name: 'reservation.month.list' })
            })

            test('Processingカバー表示終了', () => {
              expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
            })
          })
        })
      })
    })
  })
})
