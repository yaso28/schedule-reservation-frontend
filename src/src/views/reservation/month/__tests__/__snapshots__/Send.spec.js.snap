// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send マウント metaタグ 1`] = `
NodeList [
  <meta
    content="noindex"
    data-vmid="robots"
    data-vue-meta="2"
    name="robots"
  />,
]
`;

exports[`Send マウント titleタグ 1`] = `"やそスポーツクラブ | メール送信"`;

exports[`Send マウント データ取得：成功 ページ全体表示 1`] = `
<div>
  <h1>
    メール送信
  </h1>
   
  <form
    class=""
  >
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <dl
          class="row"
        >
          <dt
            class="col-sm-3"
          >
            
            月
          
          </dt>
           
          <dd
            class="col-sm-9"
          >
            
            2021年01月
          
          </dd>
           
          <dt
            class="col-sm-3"
          >
            
            予約ステータス
          
          </dt>
           
          <dd
            class="col-sm-9"
          >
            
            電話予約済
          
          </dd>
           
          <dt
            class="col-sm-3"
          >
            
            公開ステータス
          
          </dt>
           
          <dd
            class="col-sm-9"
          >
            
            未定
          
          </dd>
        </dl>
      </div>
    </div>
     
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="mail_to"
          >
             宛先メールアドレス 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <input
              class="form-control"
              id="mail_to"
              type="text"
            />
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
         
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="subject"
          >
             件名 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <input
              class="form-control"
              id="subject"
              type="text"
            />
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
         
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="message"
          >
             本文 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <textarea
              class="form-control"
              id="message"
              rows="40"
            />
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
         
        <ul>
          <li>
            
            宛先メールアドレス、件名、本文の設定変更は
            <a
              class=""
              href="/reservation/setting/edit"
              target="_self"
            >
              
              こちら
            
            </a>
          </li>
        </ul>
      </div>
       
      <footer
        class="card-footer"
      >
        <button
          class="btn btn-reset btn-secondary"
          type="button"
        >
          
          リセット
        
        </button>
         
        <button
          class="btn btn-primary"
          type="submit"
        >
          
          送信
        
        </button>
      </footer>
    </div>
  </form>
   
  <div>
    <div
      class="modal overflow-auto fade modal-primary"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="modal-dialog modal-dialog-centered modal-sm"
        role="document"
      >
        <div
          class="modal-content"
        >
          <header
            class="modal-header"
          >
            <h5
              class="modal-title"
            >
               送信しますか？ 
            </h5>
            <button
              aria-label="Close"
              class="close"
              type="button"
            >
              ×
            </button>
          </header>
          <div />
          <footer
            class="modal-footer"
          >
            <button
              class="btn btn-secondary"
              type="button"
            >
               Cancel 
            </button>
            <button
              class="btn btn-primary"
              type="button"
            >
               OK 
            </button>
          </footer>
        </div>
      </div>
    </div>
    <!---->
  </div>
</div>
`;
